pkg/power/power.go

```diff
@@ -274,43 +270,50 @@ func (p *PowerServer) setNodePowerState(cl *substrate.Substrate, up bool) error
 
 func (p *PowerServer) recv(ctx context.Context) error {
 	log.Info().Msg("listening for power events")
-	stream, err := p.consumer.PowerTargetChange(ctx)
+
+	if err := p.syncSelf(); err != nil {
+		return errors.Wrap(err, "failed to synchronize power status")
+	}
+
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := p.consumer.PowerTargetChange(subCtx)
 	if err != nil {
-		return errors.Wrapf(errConnectionError, "failed to connect to zbus events: %s", err)
+		return errors.Wrap(err, "failed to connect to zbus events")
 	}
 
 	for event := range stream {
 		if err := p.event(&event); err != nil {
-			log.Error().Err(err).Msg("failed to process power event")
+			return errors.Wrap(err, "failed to process power event")
 		}
 	}
 
+	// if we reach here it means stream was ended. this can only happen
+	// if and only if the steam was over and that can only be via a ctx
+	// cancel.
 	return nil
 }
 
 // start processing time events.
 func (p *PowerServer) events(ctx context.Context) error {
 	// first thing we need to make sure we are not suppose to be powered
 	// off, so we need to sync with grid
-	// 1) make sure at least one uptime was already sent
+	// make sure at least one uptime was already sent
 	_ = p.ut.Mark.Done(ctx)
-	// 2) do we need to power off
-	if err := p.syncSelf(); err != nil {
-		return errors.Wrap(err, "failed to synchronize power status")
-	}
 
 	// if the stream loop fails for any reason retry
 	// unless context was cancelled
 	for {
 		err := p.recv(ctx)
-		if err == nil {
-			return err
+		if err != nil {
+			log.Error().Err(err).Msg("failed to process power events")
 		}
 
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-time.After(10 * time.Second):
+		case <-time.After(5 * time.Second):
```

Comment:
Also wonder how we choose the numbers 10 and 5 in general for this

Reply:
There is no specific rule, the recv() function in general should never return, but if it does then it's probably a connection error to the chain, so there is no need to retry immediately, instead it should wait a little bit longer. seems i did 10 seconds before but now i just feel it's too long. The number of seconds to wait is just an arbitrary not too long, not to short wait to give the system time to recover